chapters/VK_EXT_present_timing/queries.txt

```diff
@@ -0,0 +1,441 @@
+
+== Present/Display Timing Queries
+
+Traditional game and real-time-animation applications frequently use
+ename:VK_PRESENT_MODE_FIFO_KHR so that presentable images are updated during
+the vertical blanking period of a given refresh cycle (RC) of the
+presentation engine's display.
+This avoids the visual anomaly known as tearing.
+
+However, synchronizing the presentation of images with the RC does not
+prevent all forms of visual anomalies.
+Stuttering occurs when the geometry for each presentable image isn't
+accurately positioned for when that image will be displayed.
+The geometry may appear to move too little some RCs, and too much for
+others.
+Sometimes the animation appears to freeze, when the same image is used for
+more than one RC.
+
+In order to minimize stuttering, an application needs to: 1) render
+and present images at a consistent rate that is a multiple of the
+presentation engine's refresh rate; 2) correctly position its geometry
+for when the presentable image will be displayed to the user.
+Applications can: benefit from communication of timing information with the
+presentation engine and its display.
+For example, applications can: determine information about the refresh
+rate of the display/compositor, can: specify when an image should be
+presented, and can: determine when an image was actually presented.
+This can allow the application's animation to look smooth to the user, with
+no stuttering.
+The
+ifdef::VK_EXT_present_timing+VK_GOOGLE_display_timing[]
+`VK_EXT_present_timing` and `VK_GOOGLE_display_timing` extensions allow
+endif::VK_EXT_present_timing+VK_GOOGLE_display_timing[]
+ifdef::VK_EXT_present_timing[]
+`VK_EXT_present_timing` extension allows
+endif::VK_EXT_present_timing[]
+ifdef::VK_GOOGLE_display_timing[]
+`VK_GOOGLE_display_timing` extension allows
+endif::VK_GOOGLE_display_timing[]
+an application to satisfy these needs.
+
+The presentation engine's display typically refreshes the pixels that are
+displayed to the user on a periodic basis.
+The period may be fixed or variable.
+In many cases, the presentation engine is associated with fixed refresh rate
+(FRR) display technology, with a fixed refresh rate (RR, e.g. 60Hz).
+In some cases, the presentation engine is associated with variable refresh
+rate (VRR) display technology, where each refresh cycle (RC) can vary in
+length.
+
+ifdef::VK_EXT_present_timing[]
+[open,refpage='vkGetSwapchainTimingPropertiesEXT',desc='Obtain the display timing properties of the PE\'s display',type='protos']
+--
+
+To query the presentation engine's timing properties for a given swapchain,
+call:
+
+include::{generated}/api/protos/vkGetSwapchainTimingPropertiesEXT.txt[]
+
+  * pname:device is the device associated with pname:swapchain.
+  * pname:swapchain is the swapchain to obtain timing properties for.
+  * pname:pSwapchainTimingProperties is a pointer to an instance of the
+    sname:VkSwapchainTimingPropertiesEXT structure.
+
```

Comment:
How could an application do an efficient blocking wait for present completion? Ie. without busy-waiting in a tight loop?
Research applications like mine often need to synchronize certain actions with present completion. E.g., starting playback of a sound at the time a certain image starts displaying, or starting audio capture, or programming some digital i/o interface card / serial port / to send out some digital TTL trigger signal to research equipment, or sampling some sensor, or sending a trigger packet over the network. Some official way of having a blocking wait for a present completion with a specific target "presentID" would be useful. Or maybe a VkFence that could be associated with such a present completion and signalled?

Reply:
Yes, that's what i want. Busy-waiting in the app is not good, especially not if you have to poll at a very high rate, because one needs lowest possible delay between present execution completion and responding to that. Drains battery on laptops, causes overhead, and on some systems the app may get penalized for such behaviour by reducing its scheduling priority, or even demoting it out of realtime scheduling for hogging a cpu too much, e.g., macOS, Windows iirc, or some of the schedulers on Linux. E.g., i already implement VK_GOOGLE_display timing support in my app, and i'd like to cut this 100 lines+ mess... https://github.com/kleinerm/Psychtoolbox-3/blob/master/PsychSourceGL/Source/Common/PsychVulkanCore/PsychVulkan.c#L2602 ... down to a handful lines, if we had block for presentID support in this extension.

If you only have polling and a frame is skipped, and even more so for variable refresh displays, you can end up with polling loops that run for dozens of milliseconds, burning cpu needlessly.

Reply:
@jjulianoatnv @keith-packard  Following up on this after some recent "play-time" with NVidia's recent proprietary Vulkan driver on Windows-10 and Linux:

Seems the VK_KHR_present_id and VK_KHR_present_wait extensions are the ones that implement this block-until-present-complete behavior. That's what i hoped for with my comments above.

I also saw that @keith-packard GitHub repo has some wip branch for Mesa on this. Great! Any ETA on when that will get integrated into Mesa?

However, the NVidia proprietary implementation seems to be quite broken atm., to the point of being useless. I couldn't get any spec compliant or recommended behavior on Windows-10 at all. No meaningfully accurate relationship between return from ``vkWaitForPresent()`` and actual image presentation. On Linux it didn't add anything in precision for the single-gpu case, and was totally broken for the Optimus / Prime render offload case as far as i could tell.

Another problem i realized is that the extension does not allow to query for support per VKSurface, only if the implementation in general supports it. So we have the same problem as mentioned in these comments wrt. querying support. Some windowing systems on Linux can easily support vkWaitForPresent(), e.g., DRM/KMS direct display on a DRM master or leased DRM output, and X11 via DRI3 + Present extension. But on Wayland, only some compositors implement the required presentation_feedback extension, others don't, so a per-WSI capability reporting would be needed for correct behavior. I guess an implementation of the extension will simply have to lie to clients as it stands, and claim support on WSI/Wayland even if the underlying Wayland compositor may not support it.

Reply:
FWIW I added an explicit interaction with VK_KHR_present_wait: after waiting on a present ID with `vkWaitForPresentKHR()`, the present timing results from that present are guaranteed to be available through `vkGetPastPresentationTimingEXT()`.